Add support for disabling gamepad inputs in the Input class

Raftatul opened the pull request:
Fixes: godotengine/godot-proposals/issues/12259

https://github.com/user-attachments/assets/7bd2cbcf-f487-496d-805b-8b0101d69046

- *Bugsquad edit:* Closes https://github.com/godotengine/godot-proposals/issues/8391 and supersedes https://github.com/godotengine/godot/pull/84848

Raftatul committed:
Add gamepad_enabled in Input class

Disable joy vibration when gamepad_enabled is false

Add support for disabling gamepad input via command line

Reformat code

Add docs and affect is_joy_button_pressed and get_joy_axis

Fix doc

Renamed "gamepad" to "joypad" and updated the documentation

Fix doc typo

Change command to disable joypad and fix project settings variable path

Fix ProjectSettings' doc for joypad_enabled

Calinou reviewed (changes requested):
Tested locally, it works as expected.

Note that setting this at startup will not defer controller initialization, so it won't sidestep https://github.com/godotengine/godot/issues/20566 on Windows. This is not critical to address the core need mentioned in https://github.com/godotengine/godot-proposals/issues/12259, but I thought I'd mention it regardless.

This PR also doesn't implement a command line argument for games that don't expose an in-game option to toggle gamepad support. As mentioned in https://github.com/godotengine/godot-proposals/issues/12259#issuecomment-2810815263, I'd like to have this as not all games will expose an option to disable gamepad support (especially smaller/gamejam projects).

Lastly, controller vibration is not disabled when gamepad support is disabled. This should be done by adding an early `return` to `Input::start_joy_vibration()` and `Input::stop_joy_vibration()`.

Raftatul commented:
> Tested locally, it works as expected.
> 
> Note that setting this at startup will not defer controller initialization, so it won't sidestep #20566 on Windows. This is not critical to address the core need mentioned in [godotengine/godot-proposals#12259](https://github.com/godotengine/godot-proposals/issues/12259), but I thought I'd mention it regardless.
> 
> This PR also doesn't implement a command line argument for games that don't expose an in-game option to toggle gamepad support. As mentioned in [godotengine/godot-proposals#12259 (comment)](https://github.com/godotengine/godot-proposals/issues/12259#issuecomment-2810815263), I'd like to have this as not all games will expose an option to disable gamepad support (especially smaller/gamejam projects).
> 
> Lastly, controller vibration is not disabled when gamepad support is disabled. This should be done by adding an early `return` to `Input::start_joy_vibration()` and `Input::stop_joy_vibration()`.

I’m not sure how to implement command lines yet, but I’ll look into it.

Raftatul commented:
> Tested locally, it works as expected.
> 
> Note that setting this at startup will not defer controller initialization, so it won't sidestep #20566 on Windows. This is not critical to address the core need mentioned in [godotengine/godot-proposals#12259](https://github.com/godotengine/godot-proposals/issues/12259), but I thought I'd mention it regardless.
> 
> This PR also doesn't implement a command line argument for games that don't expose an in-game option to toggle gamepad support. As mentioned in [godotengine/godot-proposals#12259 (comment)](https://github.com/godotengine/godot-proposals/issues/12259#issuecomment-2810815263), I'd like to have this as not all games will expose an option to disable gamepad support (especially smaller/gamejam projects).
> 
> Lastly, controller vibration is not disabled when gamepad support is disabled. This should be done by adding an early `return` to `Input::start_joy_vibration()` and `Input::stop_joy_vibration()`.

Okay, I implemented the ``--disable-gamepad`` command line option and added the ``gamepad_enabled`` setting under ``input_devices/controller/gamepad_enabled`` in the project settings. I'm not sure if it's the right way to do it, though.

MJacred commented:
@Raftatul: don't forget to squash your commits into 1

AThousandShips commented:
There's no need to squash until the PR is ready and approved

MJacred commented:
Personally, I also prefer squashing at the very end.  
But then you might get an approval and merge in quick succession, before you can squash it…

AThousandShips commented:
The PR will be squashed by maintainers if needed, or it can just be squashed by the author at request by maintainers, but we allow contributors to have separate commits for their own organization, but it's up to each contributor how they want to do things

But we prefer to not ask that contributors squash before the PR is ready, to avoid unnecessary work if the PR needs further changes, or gets rejected, and that's why we wait until it is ready

Raftatul reviewed (commented):
- Implement gamepad_enabled in Project Settings
- Implement command line to disable gamepad inputs
- Implement gamepad_enabled in start_joy_vibration and stop_joy_vibration

MJacred reviewed (changes requested):
For consistent naming, please use `joypad` instead of `gamepad`

_EDIT_: and add a note that this does **not** stop Godot from capturing/connecting to the joypad itself. For that, you need a per-platform implementation. See [this proposal](https://github.com/godotengine/godot-proposals/issues/8391) for information

Raftatul commented:
> For consistent naming, please use `joypad` instead of `gamepad`
> 
> _EDIT_: and add a note that this does **not** stop Godot from capturing/connecting to the joypad itself. For that, you need a per-platform implementation. See [this proposal](https://github.com/godotengine/godot-proposals/issues/8391) for information

Should I also rename the project settings category from `Controller` to `Joypad` ?

![image](https://github.com/user-attachments/assets/29b69f9f-a26e-4ef8-a6af-8ace4f0225d5)

MJacred reviewed (changes requested):
> Should I also rename the project settings category from `Controller` to `Joypad` ?

That's a really good question… While the [tutorial](https://docs.godotengine.org/en/stable/tutorials/inputs/controllers_gamepads_joysticks.html) mostly uses the term `Controller`, the API uses `Joypad`. I'd say conforming to the API is more important, so changing it to `Joypad` is the way to go

_EDIT_: in the xml files, it should be `does not` instead of `do not`

MJacred reviewed (approved):
This should be added to the docs in ProjectSettings.xml:
```
To control this at runtime, use [member Input.joypad_enabled].
```

apart from that, LGTM

Raftatul commented:
> This should be added to the docs in ProjectSettings.xml:
> 
> ```
> To control this at runtime, use [member Input.joypad_enabled].
> ```
> 
> apart from that, LGTM

Done

Nintorch reviewed (commented):
Tested on Windows 11, this PR mostly works as expected, but I have some suggestions:
- I think this PR should also handle the newly added `Input.set_joy_light()` that changes the LED light on a controller:
https://github.com/godotengine/godot/blob/78c6632eb174aabb2790975cf83e28fee065b43d/core/input/input.cpp#L1007-L1014 (EDIT: And controller motion sensors)
- If you use `Input.is_action_pressed()` to check for action presses, hold the button on your controller that activates the action and then disable the joypad input while still holding the button, the action will stay pressed.
There's an Input method called `Input::release_pressed_events()`, but it will also release the non-joypad events too, so I'm not too sure if we can use it.
I made a PR that saves the action event types ( https://github.com/godotengine/godot/pull/115099 ), maybe it's possible to use that functionality to check if an action is being pressed only because of a joypad press/axis movement, and if it is, release the action.

AdriaandeJongh commented on core/input/input.cpp:
```diff
@@ -1551,6 +1596,14 @@ bool Input::is_using_accumulated_input() {
 }
 
+void Input::set_joypad_enabled(bool p_enable) {
+	joypad_enabled = p_enable;
```
@Raftatul could you take a look at this comment?

Raftatul replied:
> @Raftatul could you take a look at this comment?

No! I haven’t received any ping from GitHub. I will make the change.

Raftatul replied:
I added @Nintorch's suggestion. 

Also should I remove the ``joypad_enabled`` check from the functions ``get_joy_name``, ``get_joy_vibration_strength``, ``get_joy_vibration_duration``, ``get_joy_vibration_remaining_duration``, ``set_joy_motion_sensors_enabled``, ``get_joy_motion_sensors_rate``, ``get_joy_motion_sensors_calibration``, ``set_joy_motion_sensors_calibration``, ``stop_joy_vibration``, ``get_joy_guid`` and ``get_joy_info`` ?

I feel like some of them are not necessary, like ``get_joy_guid`` and ``get_joy_name``.

Mickeon commented on doc/classes/ProjectSettings.xml:
```diff
@@ -1671,6 +1668,12 @@
+		<member name="input_devices/joypads/joypad_enabled" type="bool" setter="" getter="" default="true">
+			If [code]true[/code], enables joypad input events and allows the use of joypad-related functions such as [method Input.is_joy_button_pressed], [method Input.get_joy_axis], [method Input.start_joy_vibration], and [method Input.stop_joy_vibration].
+			This also affects [method Input.is_action_just_pressed], [method Input.is_action_just_released], and [method Input.is_action_pressed].
+			To control this at runtime, use [member Input.joypad_enabled].
```
It sounds appropriate to mention the corresponding command line argument... somewhere in here. See how other ProjectSettings descriptions do it.

Mickeon commented on doc/classes/Input.xml:
```diff
@@ -839,6 +754,11 @@
+		<member name="joypad_enabled" type="bool" setter="set_joypad_enabled" getter="is_joypad_enabled">
+			If [code]true[/code], enables joypad input events and allows the use of joypad-related functions such as [method Input.is_joy_button_pressed], [method Input.get_joy_axis], [method Input.start_joy_vibration], and [method Input.stop_joy_vibration].
+			See also [member ProjectSettings.input_devices/joypads/joypad_enabled].
+			[b]Note:[/b] Disabling this option does not stop Godot from capturing/connecting to the joypad itself.
```
"Does not stop Godot from capturing/connecting to the joypad itself" is too vague for whoever may be reading this. The note doesn't really explain why the reader should care about this

It could be interpreted as "_Godot always steals all input from the joypad_", "_Godot turns the lights on your joypad_", or even "_this property doesn't do anything_". I would suggest elaborating slightly, not too much.

I am to assume it means something more like this:
```suggestion
			[b]Note:[/b] Regardless of this property, Godot always detects how many joypads are connected (see [method get_connected_joypads]).
```
Which is admittedly, in my opinion, a bit weird. But who am I to say?

AdriaandeJongh reviewed (commented):
Just tested this PR and found that if I disable joypad input while pushing a joystick to the side, then `Input.get_action_strength()` will keep returning the last value of the action instead of the expected value of `0`.

Other than that, disabling joypads stops any events from propagating as far as I could see. Great!

Raftatul committed:
Fixed gamepad action strength not being released

Raftatul commented:
> Just tested this PR and found that if I disable joypad input while pushing a joystick to the side, then `Input.get_action_strength()` will keep returning the last value of the action instead of the expected value of `0`.
> 
> Other than that, disabling joypads stops any events from propagating as far as I could see. Great!

To fix the `Input.get_action_strength()` bug, I'll need to modify `release_pressed_events`. This code is avoiding the input to be correctly released:
```
void Input::release_pressed_events() {
	// Don't release the events if the application (or the window it's embedded in) is still focused.
	if (application_focused || embedder_focused) {
		return;
	}
```
I'm not sure why it's was added in the first place, I'll remove it for now. @Nintorch can you explain why did you added this? Was it to avoid some issues related to the controller?

Nintorch commented:
Yes, if I remember correctly it was done so that controller events aren't released accidentally when the user switches focus between the editor (the "embedder") and the embedded game window

Raftatul commented:
> Yes, if I remember correctly it was done so that controller events aren't released accidentally when the user switches focus between the editor (the "embedder") and the embedded game window

So it's only an issue related to the editor? Is it still relevant? I could bypass the check by adding a bool parameter called 'p_force_released' or something like that if we should keep it.

Raftatul committed:
Added support to ignore specific joypad device

Raftatul commented:
> I would love to review this PR if @Nintorch's comments are addressed.
> 
> One nitpick is that `joypad_enabled` refers to joypad as singular even though technically this API disables all joypad**s** (multiple). But this is not a hill I'd die on...

I pushed something to disable a specific device input. It should work, but I can’t confirm it at 100% since I don’t have several controllers available. I'll also convert this PR to draft cause I think it's not completed yet.

AdriaandeJongh commented:
> I pushed something to disable a specific device input.

That could also be part of a separate PR, as to not delay the merging of this PR any further.
